src/MBrace.Azure.Management/Compute.fs

```diff
@@ -201,20 +203,58 @@ module internal Compute =
     </Role>
 </ServiceConfiguration>""" serviceName instances storageAccount.ConnectionString serviceBusAccount.ConnectionString (if useDiagnostics then storageAccount.ConnectionString else "")
 
+    type PackageDetails = Official of mbraceVersion:string * VMSize | CustomRemote of Uri | CustomLocal of FileInfo
+    type PackageSource = Remote of Uri * Uri | Local of FileInfo
+
+    let deployPackage packageDetails destinationStorageAccount (logger:ISystemLogger) = async {
+        let! container = getDeploymentContainer destinationStorageAccount
+        let! packageSource =
+            async {
+                match packageDetails with
+                | Official (mbraceVersion, vmSize) ->
+                    // GitHub redirects to AWS for actual resources, so get the real uri of the file
+                    let originalUri = Common.getPackageUrl mbraceVersion vmSize |> Uri
+                    let webRequest = HttpWebRequest.CreateHttp originalUri
+                    webRequest.AllowAutoRedirect <- false
+                    let! response = webRequest.GetResponseAsync() |> Async.AwaitTaskCorrect
+                    return Remote(originalUri, response.Headers.["Location"] |> Uri)
+                | CustomRemote uri -> return Remote(uri, uri)
+                | CustomLocal fileInfo -> return (Local fileInfo)
+            }
+        
+        match packageSource with
+        | Remote (namedUri, resourceUri) ->       
+            let blobName =
+                let uriHash = getTextHash (namedUri.ToString())
+                sprintf "%s-%s" (Path.GetFileName namedUri.LocalPath) uriHash
+            let packageBlob = container.GetBlockBlobReference blobName
+            let! blobExists = packageBlob.ExistsAsync() |> Async.AwaitTaskCorrect
+            if not blobExists then
+                logger.Logf LogLevel.Info "deploying cloud service package package from %A to your storage account (%O)" namedUri packageBlob.Uri
+                do! packageBlob.StartCopyFromBlobAsync resourceUri |> Async.AwaitTaskCorrect |> Async.Ignore
+                while (packageBlob.CopyState.Status = CopyStatus.Pending) do
+                    do! Async.Sleep 1000
+                    do! packageBlob.FetchAttributesAsync() |> Async.AwaitTaskCorrect
+                if packageBlob.CopyState.Status <> CopyStatus.Success then
+                    return invalidOp <| sprintf "error: Unable to deploy MBrace Cloud Service package: %s" packageBlob.CopyState.StatusDescription
+            return packageBlob.Uri
+        | Local file -> 
+            let blobName =
```

Comment:
shouldn't the blob name for local files also contain a hash of it contents? Compare against line 210 in the original implementation where the `getFileHash` function is used. This is needed since a local file can often change (after a new build for instance), unlike remote urls which are less likely to be modified.